File: Diagrammatize.Design.Pattern/code/cxx/Interpreter/A1/language/InterpreterFacade.cpp
#include "InterpreterFacade.hpp"
#include "ProgramNode.hpp"
#include "ParseException.hpp"
#include "ExecuteException.hpp"

InterpreterFacade::InterpreterFacade(ExecutorFactory* factory) 
    : factory_(factory) {
}
    
bool InterpreterFacade::parse(const std::string& text) {
    bool ok = true;

    // 创建 Context 并设置 factory
    context_ = std::make_unique<Context>(text);
    context_->setExecutorFactory(factory_);

    // 创建 ProgramNode
    programNode_ = std::make_unique<ProgramNode>();

    try {
        programNode_->parse(*context_);
        std::cout << programNode_->toString() << std::endl;
    } catch (const ParseException& e) {
        std::cerr << e.what() << std::endl;
        ok = false;
    }

    return ok;
}

void InterpreterFacade::execute() {
    try {
        if (programNode_) {
            programNode_->execute();
        }
    } catch (const ExecuteException& e) {
        std::cerr << e.what() << std::endl;
    }
}

File: Diagrammatize.Design.Patterns/code/python/Interpreter/A1/language/InterpreterFacade.py

from .Executor import Executor
from .ProgramNode import ProgramNode
from .ParseException import ParseException
from .ExecuteException import ExecuteException
from .Context import Context

class InterpreterFacade(Executor):
    def __init__(self, factory):
        self._factory = factory
        self._context = None
        self._programNode = None

    def parse(self, text):
        ok = True
        self._context = Context(text)
        self._context.setExecutorFactory(self._factory)
        self._programNode = ProgramNode()

        try:
            self._programNode.parse(self._context)
            print(self._programNode)
        except ParseException as e:
            print(f"Parse error: {e}")
            ok = False

        return ok

    def execute(self):
        try:
            if self._programNode:
                self._programNode.execute()
        except ExecuteException as e:
            print(f"Execute error: {e}")
